lib/Scheduler/sched_region.cpp

```diff
@@ -77,54 +74,49 @@ void SchedRegion::CmputAbslutUprBound_() {
+// which may cause an issue when heuristic is disabled.
+FUNC_RESULT SchedRegion::FindOptimalSchedule(
     Milliseconds rgnTimeout, Milliseconds lngthTimeout, bool &isLstOptml,
     InstCount &bestCost, InstCount &bestSchedLngth, InstCount &hurstcCost,
     InstCount &hurstcSchedLngth, InstSchedule *&bestSched, bool filterByPerp,
     const BLOCKS_TO_KEEP blocksToKeep) {
   ConstrainedScheduler *lstSchdulr = NULL;
-  ConstrainedScheduler *acoSchdulr = NULL;
+  InstSchedule *InitialSchedule = nullptr;
   InstSchedule *lstSched = NULL;
-  InstSchedule *acoSchedule = NULL;
-  llvm::opt_sched::InstCount ACOScheduleLength; //TODO: CHIPPIE: Initialization value(s)? //TODO: Chippie: What is up with the similarly-named variable in the header file? hurstcSchedLngth_; in sched_region.h, but it's passed as a parameter here. Based on the context, I'm assuming ACO's schedule length and cost should instead be added as arguments to this function?
-  InstCount ACOScheduleCost;
-  bool isACOOptimal = false; //TODO: CHIPPIE: Where should this go? Should it become an additional parameter? Why is isLstOptml a reference parameter? Or don't add it as a parameter -- just rename the existing one to "isInitialScheduleOptimal" or something? Or should they all be 3 reference parameters???
-  InstSchedule *initialSchedule = NULL;
-  bool initialScheduleOptimal = false;
-  llvm::opt_sched::InstCount initialScheduleLength; //TODO: CHIPPIE: Initialization value(s)?
-  InstCount initialScheduleCost;
+  InstSchedule *AcoSchedule = nullptr;
+  InstCount InitialScheduleLength = 0;
+  InstCount InitialScheduleCost = 0;
   FUNC_RESULT rslt = RES_SUCCESS;
   Milliseconds hurstcTime = 0;
   Milliseconds boundTime = 0;
   Milliseconds enumTime = 0;
   Milliseconds vrfyTime = 0;
-  Milliseconds acoTime = 0;
-  Milliseconds acoStart = 0;
-  Milliseconds boundStart;
+  Milliseconds AcoTime = 0;
+  Milliseconds AcoStart = 0;
 
   enumCrntSched_ = NULL;
   enumBestSched_ = NULL;
   bestSched = bestSched_ = NULL;
 
-  //TODO: CHIPPIE: Perhaps rename these to match the flag names?
-  bool heuristicSchedulerEnabled = SchedulerOptions::getInstance().GetBool("HEUR_ENABLED");
-  bool acoSchedulerEnabled = SchedulerOptions::getInstance().GetBool("ACO_ENABLED");
-  bool bbSchedulerEnabled = SchedulerOptions::getInstance().GetBool("BB_ENABLED");
+  bool AcoBeforeEnum = false;
+  bool AcoAfterEnum = false;
 
-  if (false == heuristicSchedulerEnabled && false == acoSchedulerEnabled) {
-    //Abort if ACO and heuristic algorithms are disabled.
-    cout << "TODO: Descriptive error message here saying that there must be at least one of the ACO or Heuristic scheduler enabled.\n";
-    return RES_ERROR;
+  Config &schedIni = SchedulerOptions::getInstance();
+  bool HeuristicSchedulerEnabled = schedIni.GetBool("HEUR_ENABLED");
+  bool AcoSchedulerEnabled = schedIni.GetBool("ACO_ENABLED");
+  bool BbSchedulerEnabled = schedIni.GetBool("ENUM_ENABLED");
+  if (AcoSchedulerEnabled) {
```

Comment:
So if I understand the proceeding two lines correctly, ACO can be run both before and after BB?

Reply:
I thought it would be interesting to run ACO after the B&B enumerator if the B&B enumerator couldn't find an optimal solution. So instead of ACO -> B&B Enumerator, it would be something like Heuristic scheduler -> B&B Enumerator -> ACO. Also, from my previous tests with ACO, running ACO can actually take a long time, much longer than B&B. Therefore it may be good to only use ACO as a last resort. This is just an experimental option.

Reply:
I agree that it's interesting. Having B&B as an initial schedule, I wonder if it'll perform even better now.

A good experiment I'm curious about would be to see what gives a more effective final result: using ACO as an initial schedule for B&B, or using B&B as an initial schedule for ACO?

Reply:
That seems like a very good experiment to run.